Add MFD Support for Motorola mc146818 RTC, Counter, and BBRAM

akanisetti opened the pull request:
This PR introduces support for the Motorola mc146818 RTC and its associated functionalities through a Multi-Function Device (MFD) framework. The following key changes are included:

MFD Driver: Added and enabled the motorola, mc146818 MFD driver to manage RTC read/write operations and ensure data integrity through synchronized access.
RTC Driver: Enabled MFD support in the existing RTC driver for mc146818.
Counter Driver: Integrated MFD support into the CMOS counter driver.
BBRAM Driver: Enabled access to RTC RAM via a new BBRAM driver under the MFD hierarchy.
Device Tree Updates: Modified relevant .dtsi files to support the mc146818 MFD, including RTC, Counter, and BBRAM components.

These changes collectively enable full support for the mc146818 device and its subcomponents in a modular and maintainable way.

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

akanisetti commented on drivers/bbram/bbram_mc146818.c:
```diff
@@ -0,0 +1,192 @@
+	k_spinlock_key_t key = k_spin_lock(dev_data->lock);
+
+	for (size_t i = 0; i < size; i++) {
+		if (offset < MAX_STD) {
+			if (offset >= RTC_CENT) {
+
+			/* RTC_CENT byte is used to store Century data for the
+			 * RTC time and date, so skipping read/write operation
+			 * to this byte.
+			 */
+
+				*(data + i) = std_reg_read(config, offset+1);
+			} else {
+				*(data + i) = std_reg_read(config, offset);
+			}
+		} else {
+			*(data + i) = ext_reg_read(config, offset+1);
```
The suggested method will not work as if the stating offset is RTC_CENT it starts writing from the next offset and then if the stating offset is given as RTC_CENT +1 it will overwrite the data written to RTC_CENT as per the user.
So, to make it easy for the user to read/write instead of setting a hard failure for the offset RTC_CENT this logic is used.

bjarki-andreasen reviewed (changes requested):
Please check out how the ds3231 is implemented, moving common functions of the "parent" device to include/zephyr/drivers/mfd and implementation of common code to zephyr/drivers/mfd, while keeping the bbram and rtc partial drivers isolated (not sharing data structures between one or the other.

bjarki-andreasen commented:
> > Please check out how the ds3231 is implemented, moving common functions of the "parent" device to include/zephyr/drivers/mfd and implementation of common code to zephyr/drivers/mfd, while keeping the bbram and rtc partial drivers isolated (not sharing data structures between one or the other.
> 
> Sure, I will the modifications suggested. With this approach the intension is to avoid data structures between BBRAM and RTC, but they will be two independent drivers under one MFD driver as its child nodes in DTS. Please confirm if my understanding is correct.

That is correct :) It should be possible to enable only BBRAM, and only RTC parts of the device. That means, using kconfig to describe it, you should have

```
config MFD_MC146818
        bool
        help
          common code required by bbram and rtc, which will likely be a locking mechanism and a bus transfer,
          so both RTC and BBRAM can set/get regs on the "bus/device" safely and independently. The MC146818
          may not actually need this if the registers can be and will be interacted with independently, this would be
          for a i2c or spi attached device for example

config RTC_MC146818
        bool "RTC driver part"
        select MFD_MC146818
        depends on RTC

config BBRAM_MC146818
        bool "BBRAM driver part"
        select MFD_MC146818
        depends on BBRAM
```

bjarki-andreasen commented:
> > > > Please check out how the ds3231 is implemented, moving common functions of the "parent" device to include/zephyr/drivers/mfd and implementation of common code to zephyr/drivers/mfd, while keeping the bbram and rtc partial drivers isolated (not sharing data structures between one or the other.
> > > 
> > > 
> > > Sure, I will the modifications suggested. With this approach the intension is to avoid data structures between BBRAM and RTC, but they will be two independent drivers under one MFD driver as its child nodes in DTS. Please confirm if my understanding is correct.
> > 
> > 
> > That is correct :) It should be possible to enable only BBRAM, and only RTC parts of the device. That means, using kconfig to describe it, you should have
> > ```
> > config MFD_MC146818
> >         bool
> >         help
> >           common code required by bbram and rtc, which will likely be a locking mechanism and a bus transfer,
> >           so both RTC and BBRAM can set/get regs on the "bus/device" safely and independently. The MC146818
> >           may not actually need this if the registers can be and will be interacted with independently, this would be
> >           for a i2c or spi attached device for example
> > 
> > config RTC_MC146818
> >         bool "RTC driver part"
> >         select MFD_MC146818
> >         depends on RTC
> > 
> > config BBRAM_MC146818
> >         bool "BBRAM driver part"
> >         select MFD_MC146818
> >         depends on BBRAM
> > ```
> 
> A small query regarding counter, using this method COUNTER_CMOS will also have dependency on MFD_MC146818.

Sort if, COUNTER_CMOS and RTC_MC146818 are mutually exclusive, COUNTER_CMOS is just there for legacy reasons, you could refactor it to work with MFD_MC146818 yes, but unless you have a usecase specifically for COUNTER_CMOS I would leave it alone :)

akanisetti committed:
drivers: mfd: Enabled motorola,mc146818 MFD

Enabled Motorola, mc146818 MFD, which implements RTC
read/write operations and prevents data corruption
by synchronizing these operations.

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

akanisetti committed:
drivers: rtc: Enabled MFD support for RTC mc146818

Enabled support for motorola mc146818 RTC driver.

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

akanisetti committed:
drivers: counter: Enables support for MFD in counter

Enabled support for MFD in CMOS counter.

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

akanisetti committed:
drivers: bbram: Enables bbram driver for motorola, mc146818

Enables bbram driver for motorola, mc146818 under its MFD
to access the RAM of the RTC.

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

akanisetti committed:
dts: x86: Added required modifications to support RTC MFD

Added required modifications to dtsi files to support Motorola, mc146818
MFD, Counter, RTC, and BBRAM (for supported devices).

Signed-off-by: Anisetti Avinash Krishna <anisetti.avinash.krishna@intel.com>

bjarki-andreasen reviewed (changes requested):
Fantastic work! I have a few improvements but overall the addition is great!

akanisetti commented on dts/x86/intel/alder_lake.dtsi:
```diff
@@ -404,12 +404,24 @@
-			interrupts = <8 IRQ_TYPE_LOWEST_EDGE_RISING 3>;
-			interrupt-parent = <&intc>;
-			alarms-count = <1>;
+		mfd: mfd@70 {
+			compatible = "motorola,mc146818-mfd";
+			reg = <0x70 0x01 0x71 0x01 0x72 0x01 0x73 0x01>;
+
+			rtc: counter: rtc {
+				compatible = "motorola,mc146818";
+				interrupts = <8 IRQ_TYPE_LOWEST_EDGE_RISING 3>;
+				interrupt-parent = <&intc>;
+				alarms-count = <1>;
+
+				status = "okay";
+			};
+
+			bbram: bbram {
+				compatible = "motorola,mc146818-bbram";
+				size = <241>;
+				status = "okay";
+			};
```
Updated

akanisetti commented on drivers/bbram/bbram_mc146818.c:
```diff
@@ -0,0 +1,155 @@
+	DEVICE_DT_INST_DEFINE(n, &bbram_mc146818_init, NULL,		\
+				&bbram_data_##n, &bbram_config_##n,	\
+				POST_KERNEL,				\
+				CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
```
Updated

bjarki-andreasen reviewed (dismissed):
Nice!

akanisetti commented:
Hi @yperess, could you please review the PR

akanisetti commented:
PR has been created to resolve the twister issue https://github.com/zephyrproject-rtos/zephyr/pull/93403

seov-nordic reviewed (dismissed):
Looks good!

akanisetti commented:
> > Hi @yperess, could you please review the PR
> 
> Please fix the checks, whatever review I give will be dismissed automatically even if it's just a lint check that's failing.

Hi @yperess, twister issue is resolved. Could you please review the PR?

akanisetti commented:
Hi @yperess,
Could you please review this PR.

bjarki-andreasen commented:
Added myself as assignee as @yperess has been unresponsive for over a month

akanisetti commented:
> Added myself as assignee as @yperess has been unresponsive for over a month

Thank you
